ros2/src/autorccar_costmap/src/costmap.cpp

```diff
@@ -0,0 +1,227 @@
+
+void Costmap::UpdateCostmapFlag() {
+    cnt_iter += 1;
+
+    if (cnt_iter >= cnt_limit) {
+        costmap_flag = true;
+        cnt_iter = 0;
+    }
+}
+
+void Costmap::UpdateCostmap() {
+    // Get cell indices of current pose. This will be used as the start point of the bresenham line.
+    int sx = floor(cur_pose(0, 3) / resolution) + global_center_x;  // assume (global_center_x, global_center_y) = (0,0)
+    int sy = floor(cur_pose(1, 3) / resolution) + global_center_y;
+
+    // Transform point cloud from body to world frame.
+    pcl::transformPointCloud(*cur_point_cloud, *cur_point_cloud_transformed, cur_pose);
+
+    // Filtering
+    pcl::PassThrough<pcl::PointXYZI> zFilter;
+    zFilter.setInputCloud(cur_point_cloud_transformed);
+    zFilter.setFilterFieldName("z");
+    zFilter.setFilterLimits(0.0, 1.0);
+    zFilter.filter(*cur_point_cloud_filtered);
+
+    for (const auto& pt : *cur_point_cloud_filtered) {
+        // Get cell indices of the pointcloud
+        int x = floor(pt.x / resolution) + global_center_x;
+        int y = floor(pt.y / resolution) + global_center_y;
+
+        // The cell where the lidar point located is occupied.
+        UpdateCell(x, y, P_OCCUPIED);
+
+        // The cells on the bresenham line are free.
+        BresenhamLine(sx, sy, x, y);
+    }
+
+    costmap_flag = false;
+}
+
+void Costmap::BresenhamLine(int sx, int sy, int ex, int ey) {
+    int dx = std::abs(ex - sx);
+    int dy = std::abs(ey - sy);
+
+    // discriminant
+    int d;
+
+    // current point
+    int x = sx;
+    int y = sy;
+
+    std::vector<std::pair<int, int>> line;
+
+    if (dx > dy) {
+        // when 0 < |slope| < 1
+        d = 2 * dy - dx;
+        for (int i = 0; i < dx; i++) {
+            if (d < 0) {
+                d = d + 2 * dy;
+            } else {
+                if (ey > sy) {
+                    y = y + 1;
+                } else {
+                    y = y - 1;
+                }
+                d = d + 2 * (dy - dx);
+            }
+            line.push_back(std::make_pair(x, y));
+
+            if (ex > sx) {
+                x = x + 1;
+            } else {
+                x = x - 1;
+            }
+        }
+    } else {
+        // when |slope| > 1
+        d = 2 * dx - dy;
+        for (int i = 0; i < dy; i++) {
+            if (d < 0) {
+                d = d + 2 * dx;
+            } else {
+                if (ex > sx) {
+                    x = x + 1;
+                } else {
+                    x = x - 1;
+                }
+                d = d + 2 * (dx - dy);
+            }
+            line.push_back(std::make_pair(x, y));
+
+            if (ey > sy) {
+                y = y + 1;
+            } else {
+                y = y - 1;
+            }
+        }
+    }
+
+    // Set costmap.
+    for (unsigned int i = 0; i < line.size(); i++) {
+        UpdateCell(line[i].first, line[i].second, P_FREE);
+    }
+}
+
+bool Costmap::IsInBounds(int x, int y) {
```

Comment:
매우 강력 권장)
```suggestion
bool Costmap::IsInBounds(int x, int y) const {
```
const 변수와 마찬가지로
멤버변수를 바꾸지 않는 함수에는 const를 붙이는것이 좋습니다.